Mdr upgrade scenario

shylesh opened the pull request:
This PR addresses the following
1. Upgrade of clusters in MDR scenario
2. DR operator upgrades: MCO and DR hub operator
3. ACM operator upgrade
4. Pytest changes for accommodating Multicluster upgrade scenarios

shylesh committed:
MDR upgrade scenario implementation

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh committed:
MDR upgrade scenario

1. Pytest hooks for handling upgrade tests

2. Segregation of classes for handling different multicluster upgrades

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

petr-balogh commented on ocs_ci/utility/multicluster.py:
```diff
@@ -0,0 +1,212 @@
+
+
+def get_multicluster_upgrade_parametrizer():
+    return multicluster_upgrade_parametrizer["metro-dr"]()
```
We need to take this dynamically

shylesh replied:
done

petr-balogh commented on tests/conftest.py:
```diff
@@ -280,6 +285,29 @@ def export_squad_marker_to_csv(items, filename=None):
+
+    Args:
+        metafunc (pytest fixture): metafunc pytest object to access info about
+                                    test function and its parameters
```
just one tab indention.

shylesh replied:
done

shylesh commented:
Rebased on latest master

shylesh committed:
address review comments from Petr

decorate tests with role markers

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh commented on ocs_ci/utility/multicluster.py:
```diff
@@ -0,0 +1,212 @@
+                # Only option left is secondary odf
+                self.roles_to_config_index_map["Secondary_odf"] = cluster_index
+
+    def get_mdr_roles(self):
```
done

shylesh commented:
Rebased on latest master

shylesh committed:
Update roles markers

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh committed:
Add functions to choose the right parameter tuples based on roles

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh commented:
Rebased on latest master

shylesh commented on tests/conftest.py:
```diff
@@ -382,6 +410,41 @@ def pytest_collection_modifyitems(session, config, items):
+                        (MultiClusterUpgradeParametrize.MULTICLUSTER_UPGRADE_MARKERS)
+                    )
+                )
+                and ocsci_config.multicluster_scenario
```
change this to `ocsci_config.multicluster`

shylesh committed:
Add MCO and DR HUB operator upgrades class

Introduce markers for mco, drhub and acm upgrades

Add tests for mco and drhub upgrades

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh commented on ocs_ci/ocs/dr_upgrade.py:
```diff
@@ -0,0 +1,134 @@
+        self.pre_upgrade_images = self.get_pre_upgrade_image(self.csv_name_pre_upgrade)
+        self.load_version_config_file(self.upgrade_version)
+
+        with CephHealthMonitor(self.ceph_cluster):
```
Need to instantiate self.ceph_cluster

petr-balogh replied:
Suggest dropping out the ceph health monitor for now, as this will require more adjustment in the ceph cluster object and also in the health monitor context manager to enforce it to run on specific cluster and propagate specific kubeconfig for all of underlaying calls.

shylesh commented:
Rebased on latest master

petr-balogh commented on ocs_ci/utility/multicluster.py:
```diff
@@ -0,0 +1,238 @@
+        # This rank comes handy when we have to order the tests
+        self.zone_ranks = {}
+
+    def get_roles(self, metafunc):
```
Add `@abstractmethod` decorator to throw exception if not implemented.

petr-balogh commented on ocs_ci/utility/multicluster.py:
```diff
@@ -0,0 +1,238 @@
+                self.zone_base_rank + i * self.zone_base_rank
+            )
+
+    def generate_role_ranks(self):
```
Add `@abstractmethod` decorator to throw exception if not implemented.

petr-balogh commented on ocs_ci/utility/multicluster.py:
```diff
@@ -0,0 +1,238 @@
+        This will be called for every testcase parametrization
+
+        """
+        pass
```
Add `@abstractmethod` decorator to throw exception if not implemented.

shylesh committed:
Add ACM upgrade class

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

petr-balogh commented on ocs_ci/utility/multicluster.py:
```diff
@@ -0,0 +1,238 @@
+    def generate_config_index_map(self):
+        """
+        Generate config indexes for all the MDRs cluster roles
+        ex: {"ActiveACM": 0, "PassiceACM": 2, "PrimaryODF": 1, "SecondaryODF": 3}
```
```suggestion
        ex: {"ActiveACM": 0, "PassiveACM": 2, "PrimaryODF": 1, "SecondaryODF": 3}
```
Change all occurrences below as well

shylesh committed:
Address review comments

1.Introduce abstract methods

2. Fix typos

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh commented:
Rebased on latest master

shylesh commented:
Rebased on latest master

shylesh committed:
Add validations for MCO, DR and ACM upgrades

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh commented:
Rebased on latest master

shylesh committed:
Add facility for handling acm upgrade versions

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh commented:
Rebased on latest master

shylesh commented:
Rebased on latest master

shylesh committed:
Add skipif_z_stream markers to DR hub and DR cluster operator upgrade tests

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh commented:
Rebased on latest master

shylesh commented:
Rebased on latest master

petr-balogh commented on ocs_ci/ocs/dr_upgrade.py:
```diff
@@ -0,0 +1,271 @@
+        self.operator_name = defaults.MCO_OPERATOR_NAME
+        self.subscription_name = constants.MCO_SUBSCRIPTION
+
+    def run_upgrade(self):
```
In the case of Z-stream upgrade. We should validate upgrade of DR components in ODF upgrade test case which means.

Before we update catalogSource in ODF upgrade test - we should add code to check if it's Multicluster DR upgrade, if so, init the object of this class, run collect_data and continue with ODF upgrade as usual, once we finish ODF upgrade -> run validation which depends on collected date before upgrade.

shylesh commented:
Rebased on latest master

shylesh committed:
Fix test ordering issues

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh commented:
Rebased on latest master

shylesh committed:
Handle CephCluster object in case of ACM cluster context during upgrade

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh commented:
Rebased on the latest master

shylesh committed:
Add dummy cephcluster classes for healthmonitor

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh commented:
Rebased on the latest master

shylesh committed:
Fix minor zstream upgrade issues

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh commented:
Rebased and resolved the conflicts

shylesh committed:
Make adjustments for the new order marker

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh committed:
Minor fixes for test ordering

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh commented:
Rebased on the latest master

shylesh commented:
Rebased on the latest master

shylesh commented:
Rebased on the latest master

shylesh committed:
Fix few issues in upgrades

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh commented:
Rebased on the latest master

shylesh commented:
Rebased on the latest master

shylesh commented:
Rebased on the latest master

shylesh commented:
Rebased on the latest master/resolved conflicts.

shylesh commented:
Rebased on the latest master

shylesh committed:
Add validation for dr operators as part of ocs upgrade test in the case of zstream upgrade

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh commented:
Rebased on latest master and conflicts have been resolved

shylesh committed:
Add --acm-version CLI option

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh commented:
Rebased on the latest master

shylesh commented:
Rebased on the latest master

shylesh commented:
Rebased on the latest master

shylesh commented:
Rebased on the latest master

shylesh committed:
Handle reload of the configs during upgrades

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh commented:
Rebased on the latest master

shylesh commented:
Rebased on the latest master

shylesh committed:
Save Preupgrade conf values in PREUPGRADE_CONFIG attr of the Config class

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh commented:
Rebased on the latest master

shylesh committed:
Fix subscription naming issue

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh commented:
Rebased on the latest master

shylesh commented:
Rebased on the latest master

shylesh committed:
Rely on CSV list to find pre-upgrade csvs rather than PackageManifest

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

petr-balogh commented on ocs_ci/framework/conf/default_config.yaml:
```diff
@@ -349,3 +349,7 @@ MULTICLUSTER:
+
+PREUPGRADE_CONFIG:
+  AUTH: null 
+  MULTICLUSTER: null 
```
missing empty line at the end of file

petr-balogh commented on ocs_ci/ocs/ocs_upgrade.py:
```diff
@@ -358,23 +363,29 @@ def load_version_config_file(self, upgrade_version):
-        channel = config.DEPLOYMENT.get("ocs_csv_channel")
 
-        return package_manifest.get_current_csv(channel)
+        csv_name = None
+        csv_list = get_csvs_start_with_prefix(resource_name, namespace=self.namespace)
+        for csv in csv_list:
+            if resource_name in csv.get("metadata").get("name"):
+                if config.PREUPGRADE_CONFIG.get("ENV_DATA").get(
+                    "ocs_version"
+                ) in csv.get("metadata").get("name"):
+                    csv_name = csv.get("metadata").get("name")
+        if not csv_name:
+            raise CSVNotFound(f"No preupgrade CSV found for {resource_name}")
+        return csv_name
```
```suggestion
        csv_list = get_csvs_start_with_prefix(resource_name, namespace=self.namespace)
        for csv in csv_list:
            if resource_name in csv.get("metadata").get("name"):
                if config.PREUPGRADE_CONFIG.get("ENV_DATA").get(
                    "ocs_version"
                ) in csv.get("metadata").get("name"):
                    return csv.get("metadata").get("name")
        raise CSVNotFound(f"No preupgrade CSV found for {resource_name}")
```

shylesh committed:
Fix acm issues:
1. Catsrc name patching issue
2. Format icsp yaml
3. patch channel name

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh commented:
rebased on the latest master

shylesh commented:
Rebased on latest master

shylesh committed:
Sleep for 10 seconds after updating the subscription

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh committed:
Collect upgrade version in the init

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh committed:
Add pre and post upgrade CSV version key check

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh committed:
implement check_if_upgrade_completed() function for DR class

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh committed:
Fix tox issues for black

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh committed:
Address review comments from Petr

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh commented:
rebased on the latest master

dahorak commented on tests/conftest.py:
```diff
@@ -445,6 +475,63 @@ def pytest_collection_modifyitems(session, config, items):
+                        (MultiClusterUpgradeParametrize.MULTICLUSTER_UPGRADE_MARKERS)
+                    )
+                )
+                and ocsci_config.multicluster
```
this part of the condition `and ocsci_config.multicluster` is probably not necessary here as it is also part of the condition on line 479:
```python
if ocsci_config.multicluster and ocsci_config.UPGRADE.get("upgrade", False):
```

dahorak reviewed (dismissed):
I have just two questions mentioned above (but they are more questions or notes, not thinks which have to be necessarily fixed).
I don't have so deep knowledge in the existing  upgrade automation, so I'm not able to easily distinguish some possible technical gaps if there are any, so overall LGTM.

shylesh committed:
Handle default values for DR params in test function for non-DR cases

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

shylesh commented:
Latest runs:
verification: https://url.corp.redhat.com/finalrun
Regression: https://url.corp.redhat.com/regression-final

shylesh commented:
/cherry-pick release-4.17
